tests/test_save.py

```diff
@@ -27,14 +27,14 @@ def run_simulation(save=False, dx=0.01, dy=0.01, a=0.5, timesteps=100):
 
     grid = Grid(shape=(nx, ny))
     u = TimeFunction(
-        name='u', grid=grid, time_dim=timesteps, initializer=initializer,
-        time_order=1, space_order=2, save=save
+        name='u', grid=grid, save=timesteps, initializer=initializer,
+        time_order=1, space_order=2
     )
 
     eqn = Eq(u.dt, a * (u.dx2 + u.dy2))
     stencil = solve(eqn, u.forward)[0]
     op = Operator(Eq(u.forward, stencil), time_axis=Forward)
-    op.apply(time=timesteps, dt=dt)
+    op.apply(time=timesteps-1, dt=dt)
```

Comment:
This is odd. :thinking: What changed to cause this, and why do you change it?

Reply:
I'll be honest, I don't understand why this was required, but the test was failing without it. Maybe I'll spend more time.

Reply:
Yes, it would be good to understand why this is necessary. What happens without it?

Reply:
The bug highlighted in https://github.com/opesci/devito/pull/426 is triggered, i.e. it seems that we are trying to run for more timesteps than the size of the time dimension of the function with `save=True`.

Reply:
Ok, to be clear. The symptom of the bug in #426 is a segfault on my macbook. When I tried running this test on master, it also segfaulted and making this change made the segfault go away. 
My hypothesis now is that this test was only run on machines with more relaxed memory-access policies and that this test was always "broken" in that sense. 
Here's my best-guess for reasons: prior to the introduction of checkpointing, calling `Operator.apply(nt)` would run the operator for `nt-time_order` timesteps so allocating `nt` elements in `TimeFunction`s along the time dimension would work fine with no illegal memory accesses. During the checkpointing work, I changed the behaviour so `Operator.apply(nt)` runs `nt` timesteps to make things more transparent outside the API. I think the cleanest way to handle this would be to allocate `save + time_order` elements in the time dimension but I'm wary of allocating more memory than the user expects, especially since we will typically be running in memory-constrained environments.

Reply:
> I'm wary of allocating more memory than the user expect

I agree this is not on the table. Good spot for the rest.

Reply:
I find this answer wholly unsatisfying. #426 forces an explicit overrun in the time dimension, while this case merely tries to use the same size as used for allocation. If this is deemed "wrong" and a bug, we probably have a long list of "bugs" in the test base, since this is pretty much default behaviour, no? Simply glossing over that fact with a simple `-1` sets a dangerous precedent, especially if we are chasing sporadic bugs that may well be due to memory corruption.